File: instagram-php-scraper-master/scraper/get_user_ids.php
<?php
// get_user_ids.php
// Usage: php get_user_ids.php username1 username2 username3 ...
// Or: php get_user_ids.php --file usernames.txt
// Env required: IG_SESSIONID, IG_CSRF, IG_DS_USER_ID
// Optional: IG_UA
declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Handler\CurlHandler;
use GuzzleHttp\HandlerStack;

/* ---------- helpers ---------- */
function envs(string $k, ?string $def=''): string { 
    $v=getenv($k); 
    return ($v===false)?(string)$def:(string)$v; 
}

function jdie(string $m, int $c=1): void { 
    fwrite(STDERR, $m.PHP_EOL); 
    exit($c); 
}

/* ---------- inputs/env ---------- */
$usernames = [];

// Check if --file flag is used
if (isset($argv[1]) && $argv[1] === '--file') {
    if (!isset($argv[2])) {
        jdie("Usage: php get_user_ids.php --file <filename>\n   or: php get_user_ids.php username1 username2 ...", 2);
    }
    
    $filename = $argv[2];
    if (!file_exists($filename)) {
        jdie("File not found: {$filename}", 3);
    }
    
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    if ($lines === false) {
        jdie("Could not read file: {$filename}", 3);
    }
    
    foreach ($lines as $line) {
        $line = trim($line);
        // Skip comments and empty lines
        if ($line === '' || $line[0] === '#') continue;
        $usernames[] = $line;
    }
} else {
    // Get usernames from command line arguments
    if (count($argv) < 2) {
        jdie("Usage: php get_user_ids.php username1 username2 ...\n   or: php get_user_ids.php --file usernames.txt", 2);
    }
    
    $usernames = array_slice($argv, 1);
}

if (empty($usernames)) {
    jdie("No usernames provided", 2);
}

// Get Instagram credentials from environment
$csrf = envs('IG_CSRF', null);
$sess = envs('IG_SESSIONID', null);
$dsid = envs('IG_DS_USER_ID', null);
$ua   = envs('IG_UA', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36');

if (!$csrf || !$sess || !$dsid) {
    jdie("Missing env: IG_CSRF / IG_SESSIONID / IG_DS_USER_ID", 3);
}

/* ---------- setup HTTP client ---------- */
$handler = HandlerStack::create(new CurlHandler());
$verifyPath = ini_get('curl.cainfo');
if (!$verifyPath) $verifyPath = ini_get('openssl.cafile') ?: true;

$client = new Client([
    'base_uri' => 'https://www.instagram.com/',
    'handler' => $handler,
    'http_errors' => false,
    'timeout' => 30,
    'decode_content' => true,
    'verify' => $verifyPath,
    'force_ip_resolve' => 'v4',
    'curl' => [CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1],
    'headers' => [
        'User-Agent' => $ua,
        'Referer' => 'https://www.instagram.com/',
        'Origin' => 'https://www.instagram.com',
        'Accept' => '*/*',
        'Accept-Language' => 'en-US,en;q=0.9',
        'Accept-Encoding' => 'gzip, deflate',
        'X-Requested-With' => 'XMLHttpRequest',
        'X-IG-App-ID' => '936619743392459',
        'X-CSRFToken' => $csrf,
        'Cookie' => "csrftoken={$csrf}; sessionid={$sess}; ds_user_id={$dsid};",
        'Sec-Fetch-Site' => 'same-origin',
        'Sec-Fetch-Mode' => 'cors',
        'Sec-Fetch-Dest' => 'empty',
    ],
]);

/* ---------- fetch user IDs ---------- */
$results = [];
$errors = [];

echo "Fetching user IDs for " . count($usernames) . " username(s)...\n\n";

foreach ($usernames as $username) {
    $username = trim($username);
    if ($username === '') continue;
    
    echo "Processing: @{$username} ... ";
    
    try {
        // Use Instagram's web API to get user info
        $res = $client->get("api/v1/users/web_profile_info/?username=" . urlencode($username));
        $code = $res->getStatusCode();
        $body = (string)$res->getBody();
        
        if ($code !== 200) {
            echo "ERROR (HTTP {$code})\n";
            $errors[] = [
                'username' => $username,
                'error' => "HTTP {$code}",
                'body' => substr($body, 0, 200)
            ];
            continue;
        }
        
        $data = json_decode($body, true);
        
        if (!is_array($data)) {
            echo "ERROR (Invalid JSON)\n";
            $errors[] = [
                'username' => $username,
                'error' => 'Invalid JSON response'
            ];
            continue;
        }
        
        // Extract user ID from response
        $userId = $data['data']['user']['id'] ?? null;
        
        if (!$userId) {
            echo "ERROR (User ID not found)\n";
            $errors[] = [
                'username' => $username,
                'error' => 'User ID not found in response'
            ];
            continue;
        }
        
        $fullName = $data['data']['user']['full_name'] ?? '';
        $isPrivate = $data['data']['user']['is_private'] ?? false;
        $followerCount = $data['data']['user']['edge_followed_by']['count'] ?? 0;
        
        $results[] = [
            'username' => $username,
            'user_id' => $userId,
            'full_name' => $fullName,
            'is_private' => $isPrivate,
            'followers' => $followerCount
        ];
        
        echo "✓ ID: {$userId}\n";
        
        // Small delay to avoid rate limiting
        usleep(500000); // 0.5 seconds
        
    } catch (\Throwable $e) {
        echo "ERROR (" . $e->getMessage() . ")\n";
        $errors[] = [
            'username' => $username,
            'error' => $e->getMessage()
        ];
    }
}

/* ---------- output results ---------- */
echo "\n" . str_repeat("=", 60) . "\n";
echo "RESULTS\n";
echo str_repeat("=", 60) . "\n\n";

if (!empty($results)) {
    echo "Successfully fetched " . count($results) . " user ID(s):\n\n";
    
    // Table format
    printf("%-20s %-15s %-30s %s\n", "Username", "User ID", "Full Name", "Followers");
    echo str_repeat("-", 100) . "\n";
    
    foreach ($results as $r) {
        printf(
            "%-20s %-15s %-30s %s%s\n",
            "@" . $r['username'],
            $r['user_id'],
            mb_substr($r['full_name'], 0, 28),
            number_format($r['followers']),
            $r['is_private'] ? ' 🔒' : ''
        );
    }
    
    echo "\n" . str_repeat("-", 60) . "\n";
    echo "User IDs only (for batch processing):\n";
    echo implode("\n", array_column($results, 'user_id')) . "\n";
    
    echo "\n" . str_repeat("-", 60) . "\n";
    echo "Command to run stories_with_stickers.php for each user:\n\n";
    
    foreach ($results as $r) {
        echo "php stories_with_stickers.php {$r['user_id']}  # @{$r['username']}\n";
    }
    
    // Save results to JSON file
    $outputFile = __DIR__ . '/user_ids_' . date('Y-m-d_His') . '.json';
    file_put_contents($outputFile, json_encode([
        'timestamp' => date('c'),
        'total_requested' => count($usernames),
        'total_found' => count($results),
        'total_errors' => count($errors),
        'results' => $results,
        'errors' => $errors
    ], JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
    
    echo "\n✓ Results saved to: {$outputFile}\n";
}

if (!empty($errors)) {
    echo "\n" . str_repeat("=", 60) . "\n";
    echo "ERRORS (" . count($errors) . ")\n";
    echo str_repeat("=", 60) . "\n\n";
    
    foreach ($errors as $err) {
        echo "• @{$err['username']}: {$err['error']}\n";
    }
}

echo "\n";

File: instagram-php-scraper-master/analyzer/get_user_ids.py
#!/usr/bin/env python3
"""
get_user_ids.py - Python port of scraper/get_user_ids.php.

Converts a list of Instagram usernames to user_ids via the private
web_profile_info endpoint, and writes user_ids_<timestamp>.json to the
project root - the same location run_daily_stories.py's
find_latest_json_file() already reads from.

Usage:
    python get_user_ids.py username1 username2 ...
    python get_user_ids.py --file influencers.txt
"""

import argparse
import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List

from ig_scraper.client import MissingSessionError, fetch_user_profile, fetch_user_profile_html

if sys.stdout.encoding and sys.stdout.encoding.lower() != "utf-8":
    try:
        sys.stdout.reconfigure(encoding="utf-8")
        sys.stderr.reconfigure(encoding="utf-8")
    except Exception:
        pass

PROJECT_ROOT = Path(__file__).resolve().parent.parent


def load_usernames(args: argparse.Namespace) -> List[str]:
    if args.file:
        path = Path(args.file)
        if not path.is_file():
            print(f"File not found: {args.file}", file=sys.stderr)
            sys.exit(3)
        lines = path.read_text(encoding="utf-8").splitlines()
        return [ln.strip() for ln in lines if ln.strip() and not ln.strip().startswith("#")]
    if args.usernames:
        return [u.strip() for u in args.usernames if u.strip()]
    print("Usage: python get_user_ids.py username1 username2 ...\n   or: python get_user_ids.py --file usernames.txt", file=sys.stderr)
    sys.exit(2)


def main() -> None:
    parser = argparse.ArgumentParser(description="Resolve Instagram usernames to user_ids")
    parser.add_argument("usernames", nargs="*", help="usernames to resolve")
    parser.add_argument("--file", help="path to a file with one username per line")
    args = parser.parse_args()

    usernames = load_usernames(args)
    print(f"Fetching user IDs for {len(usernames)} username(s)...\n")

    results: List[Dict[str, Any]] = []
    errors: List[Dict[str, Any]] = []

    for username in usernames:
        print(f"Processing: @{username} ... ", end="")
        try:
            try:
                data = fetch_user_profile(username)
            except MissingSessionError:
                raise
            except Exception as api_error:
                print(f"[API failed: {api_error}] falling back to HTML scrape... ", end="")
                data = fetch_user_profile_html(username)
            user = (data.get("data") or {}).get("user") or {}
            user_id = user.get("id")
            if not user_id:
                print("ERROR (User ID not found)")
                errors.append({"username": username, "error": "User ID not found in response"})
                continue
            results.append({
                "username": username,
                "user_id": user_id,
                "full_name": user.get("full_name", ""),
                "is_private": bool(user.get("is_private", False)),
                "followers": ((user.get("edge_followed_by") or {}).get("count", 0)),
            })
            print(f"✓ ID: {user_id}")
        except MissingSessionError as e:
            print(f"ERROR ({e})")
            sys.exit(3)
        except Exception as e:  # noqa: BLE001 - mirrors PHP's catch-all + continue
            print(f"ERROR ({e})")
            errors.append({"username": username, "error": str(e)})

        time.sleep(0.5)  # mirrors get_user_ids.php's usleep(500000)

    print("\n" + "=" * 60)
    print("RESULTS")
    print("=" * 60 + "\n")

    if results:
        print(f"Successfully fetched {len(results)} user ID(s):\n")
        print(f"{'Username':<20} {'User ID':<15} {'Full Name':<30} Followers")
        print("-" * 100)
        for r in results:
            privacy = " \U0001F512" if r["is_private"] else ""
            print(f"@{r['username']:<19} {r['user_id']:<15} {r['full_name'][:28]:<30} {r['followers']:,}{privacy}")

        print("\n" + "-" * 60)
        print("User IDs only (for batch processing):")
        print("\n".join(str(r["user_id"]) for r in results))

        output_file = PROJECT_ROOT / f"user_ids_{datetime.now().strftime('%Y-%m-%d_%H%M%S')}.json"
        output_file.write_text(
            json.dumps(
                {
                    "timestamp": datetime.now().isoformat(),
                    "total_requested": len(usernames),
                    "total_found": len(results),
                    "total_errors": len(errors),
                    "results": results,
                    "errors": errors,
                },
                ensure_ascii=False,
                indent=2,
            ),
            encoding="utf-8",
        )
        print(f"\n✓ Results saved to: {output_file}")

    if errors:
        print("\n" + "=" * 60)
        print(f"ERRORS ({len(errors)})")
        print("=" * 60 + "\n")
        for err in errors:
            print(f"• @{err['username']}: {err['error']}")

    print()


if __name__ == "__main__":
    main()
